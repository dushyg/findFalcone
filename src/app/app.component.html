<div class="falconApp">
  <app-falcon-header
    (headerLinkClicked)="headerLinkClicked($event)"
  ></app-falcon-header>
  <div class="appFalconBody">
    <div class="errorMsg" *ngIf="finderFacadeService.error$ | async as error">
      <span>{{ error }}</span>
    </div>
    <router-outlet></router-outlet>
  </div>
  <app-falcon-footer></app-falcon-footer>
</div>
