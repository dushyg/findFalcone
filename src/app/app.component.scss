.falconApp {
  width: 100%;
  app-falcon-header {
    height: 60px;
    position: fixed;
    top: 0;
    width: 100%;
    background-color: white;
  }

  app-finder-board {
    height: 100%;
  }
  .appFalconBody {
    margin-top: 100px;
    height: 100%;

    .errorMsg {
      margin: 10px auto;
      text-align: center;
      color: red;
    }
  }

  app-falcon-footer {
    height: 40px;
  }
}
