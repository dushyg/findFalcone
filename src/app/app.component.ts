import {
  Component,
  OnInit,
  AfterViewChecked,
  ChangeDetectionStrategy,
} from '@angular/core';
import { FinderFacadeService } from './services/finder-facade.service';
import { constants } from './shared/constants';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AppComponent implements OnInit {
  title = 'findingFalcone';
  constructor(public finderFacadeService: FinderFacadeService) {}
  ngOnInit() {}

  public headerLinkClicked(linkName: string): void {
    if (linkName === constants.resetLink) {
      this.finderFacadeService.resetApp();
    }
  }
}
