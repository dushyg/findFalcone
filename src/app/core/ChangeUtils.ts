import { ISearchAttempt } from "./models/searchAttempt";
import { IPlanet } from "./models/planet";
import { IVehicle } from "./models/vehicle";
import { IFalconAppState } from "./models/falconApp.state";

export class ChangeUtils {
  public static getNextStateAfterChange(
    changedWidgetId: number,
    changedValue: string,
    changedWidgetSearchAttemptGetter: () => ISearchAttempt,
    previousState: IFalconAppState
  ): IFalconAppState {
    const updatedSearchMap = ChangeUtils.getUpdatedSearchMap(
      previousState.searchMap,
      changedWidgetId,
      changedValue,
      changedWidgetSearchAttemptGetter
    );

    // find unsearched planets after this planet change
    const unsearchedPlanets: IPlanet[] = ChangeUtils.getUnsearchedPlanets(
      updatedSearchMap,
      previousState.planetList
    );

    // update the vehicle inventory, taking into account used and freed vehicles
    const updatedVehicleInventory: IVehicle[] = ChangeUtils.getUpdatedVehicleInventory(
      updatedSearchMap,
      previousState.vehicleList
    );

    // calculate the total time taken to search planets with vehicles selected
    const totalTimeTakenForSearch: number = ChangeUtils.getTotalTimeTakenForSearch(
      updatedSearchMap,
      previousState.planetList,
      previousState.vehicleList
    );

    const isReadyToSearch: boolean = ChangeUtils.isItFineToStartSearching(
      updatedSearchMap
    );

    return <IFalconAppState>{
      ...previousState,
      searchMap: updatedSearchMap,
      unsearchedPlanets,
      vehicleInventory: updatedVehicleInventory,
      isReadyToSearch,
      totalTimeTaken: totalTimeTakenForSearch,
    };
  }

  private static getUpdatedSearchMap(
    searchMap: Map<string, ISearchAttempt>,
    changedWidgetId: number,
    changedValue: string,
    changedWidgetSearchAttemptGetter: () => ISearchAttempt
  ): Map<string, ISearchAttempt> {
    let updatedSearchMap = new Map<string, ISearchAttempt>();

    searchMap.forEach((value: ISearchAttempt, key: string) => {
      let updatedSearchAttempt: ISearchAttempt;
      // if the current widget is to the left of changed widget or the changed widget then
      // just make a clone for immutability and continue
      const currentlyLoopedWidgetId = Number(key);
      if (currentlyLoopedWidgetId < changedWidgetId) {
        updatedSearchAttempt = <ISearchAttempt>{ ...value };
      } else if (currentlyLoopedWidgetId == changedWidgetId) {
        updatedSearchAttempt = changedWidgetSearchAttemptGetter();
      } else {
        // else if the current widget is to the right of the changed widget then
        // we will need to reset both the existing planet and vehicle selections if any
        updatedSearchAttempt = <ISearchAttempt>{
          vehicleUsed: null,
          searchedPlanet: null,
        };
      }

      updatedSearchMap.set(key, updatedSearchAttempt);
    });
    return updatedSearchMap;
  }

  private static getUnsearchedPlanets(
    searchMap: Map<string, ISearchAttempt>,
    planetList: IPlanet[]
  ): IPlanet[] {
    const usedPlanetSet = new Set<string>();

    searchMap.forEach((searchAttempt: ISearchAttempt) => {
      if (searchAttempt && searchAttempt.searchedPlanet) {
        usedPlanetSet.add(searchAttempt.searchedPlanet);
      }
    });

    return planetList.filter(
      (planet: IPlanet) => !usedPlanetSet.has(planet.name)
    );
  }

  private static getUpdatedVehicleInventory(
    searchMap: Map<string, ISearchAttempt>,
    vehicleList: IVehicle[]
  ): IVehicle[] {
    // create a map of vehicle name vs count of vehicles used
    const usedVehicleMap = this.getUsedVehicleMap(searchMap);

    // update vehicle list with available vehicle units
    let updatedVehicleList: IVehicle[] = vehicleList.map((vehicle) => {
      const totalNumUnits = vehicle.totalNumUnits;
      return <IVehicle>{
        ...vehicle,
        availNumUnits: totalNumUnits - (usedVehicleMap.get(vehicle.name) || 0),
      };
    });

    return updatedVehicleList;
  }

  private static getUsedVehicleMap(
    searchMap: Map<string, ISearchAttempt>
  ): Map<string, number> {
    const usedVehicleMap = new Map<string, number>();
    for (let index = 1; index < searchMap.size + 1; index++) {
      const key = index.toString();
      const searchAttempt: ISearchAttempt = searchMap.get(key);
      if (searchAttempt) {
        const vehicle: string = searchAttempt.vehicleUsed;
        if (vehicle) {
          let count: number = usedVehicleMap.get(vehicle);
          if (Number(count)) {
            usedVehicleMap.set(vehicle, ++count);
          } else {
            usedVehicleMap.set(vehicle, 1);
          }
        }
      }
    }
    return usedVehicleMap;
  }

  private static getTotalTimeTakenForSearch(
    searchMap: Map<string, ISearchAttempt>,
    allPlanets: IPlanet[],
    allVehicles: IVehicle[]
  ): number {
    let totalTimeTakenForSearch: number = 0;

    const nameToPlanetMap: Map<string, IPlanet> = new Map<string, IPlanet>();
    const nameToVehicleMap: Map<string, IVehicle> = new Map<string, IVehicle>();

    allPlanets.forEach((planet) => {
      nameToPlanetMap.set(planet.name, planet);
    });

    allVehicles.forEach((vehicle) => {
      nameToVehicleMap.set(vehicle.name, vehicle);
    });

    searchMap.forEach((value: ISearchAttempt) => {
      if (!value) {
        return;
      }

      if (!value.searchedPlanet || !value.vehicleUsed) {
        return;
      }

      const planet: IPlanet = nameToPlanetMap.get(value.searchedPlanet);
      const vehicle: IVehicle = nameToVehicleMap.get(value.vehicleUsed);
      const planetDistance = planet.distance;
      const vehicleSpeed = vehicle.speed;
      if (!planetDistance || !vehicleSpeed) {
        return;
      }

      totalTimeTakenForSearch =
        totalTimeTakenForSearch + planetDistance / vehicleSpeed;
    });

    return totalTimeTakenForSearch;
  }

  private static isItFineToStartSearching(
    searchMap: Map<string, ISearchAttempt>
  ): boolean {
    let isReadyForSearch: boolean = true;

    searchMap.forEach((value: ISearchAttempt) => {
      if (!value) {
        isReadyForSearch = false;
        return;
      }

      if (!value.searchedPlanet || !value.vehicleUsed) {
        isReadyForSearch = false;
        return;
      }
    });

    return isReadyForSearch;
  }
}
