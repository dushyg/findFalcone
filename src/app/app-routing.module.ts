import { NgModule } from "@angular/core";
import { Routes, RouterModule, PreloadAllModules } from "@angular/router";
import { FinderBoardComponent } from "./finder-board/smartComponents/finder-board/finder-board.component";
import { FalconeResetComponent } from "./shared/presenterComponents/falcone-reset/falcone-reset.component";

const routes: Routes = [
  {
    path: "",
    redirectTo: "finderboard",
    pathMatch: "full",
  },
  {
    path: "finderboard",
    component: FinderBoardComponent,
  },
  {
    path: "result",
    loadChildren: () =>
      import("./finder-result/finderResult.module").then(
        (m) => m.FinderResultModule
      ),
  },
  {
    path: "reset",
    component: FalconeResetComponent,
  },
  {
    path: "**",
    redirectTo: "finderboard",
  },
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
})
export class AppRoutingModule {}
