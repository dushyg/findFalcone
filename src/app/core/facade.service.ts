import { Injectable } from '@angular/core';
import { PlanetsService } from './planets.service';
import { VehiclesService } from './vehicles.service';
import { FalconFinderService } from './falcon-finder.service';
import { Router } from '@angular/router';
import { IFindFalconRequest } from './models/find-falcon-request';
import { IFindFalconResponse } from './models/find-falcon-response';
import { catchError, map, withLatestFrom, tap, distinctUntilChanged } from 'rxjs/operators';
import { Subject, EMPTY, BehaviorSubject, Observable, combineLatest, forkJoin } from 'rxjs';
import PlanetChange from './models/planetChange';
import VehicleChange from './models/vehicleChange';
import PlanetUpdates from './models/planetUpdates';
import { IPlanet } from './models/planet';
import { IVehicle } from './models/vehicle';
import VehicleUpdates from './models/vehicleUpdates';
import { ISearchAttempt } from './models/searchAttempt';

@Injectable({providedIn : 'root'})
export default class FalconeFacade {
            
    constructor(private planetService : PlanetsService,
                private vehicleService : VehiclesService,
                private finderService : FalconFinderService,
                private router : Router){
        
    }

    private finderApiToken : string;       

    private searchMapSubject = new Subject<Map<number, ISearchAttempt>>();
    public searchMap$ : Observable<Map<number, ISearchAttempt>> = this.searchMapSubject.asObservable();

    private readonly MAX_SEARCH_ATTEMPTS_ALLOWED = 4; 
    private maxSearchAttemptAllowedSubject = new BehaviorSubject<number>(this.MAX_SEARCH_ATTEMPTS_ALLOWED);    
    public maxSearchAttemptAllowed$ : Observable<number> = this.maxSearchAttemptAllowedSubject.asObservable();
    public getMaxSearchAttemptsAllowedCount() {
        return this.MAX_SEARCH_ATTEMPTS_ALLOWED;
    }
    private errorMessageSubject = new Subject<string>();
    public errorMessage$ : Observable<string> = this.errorMessageSubject.asObservable();

    private apiVehiclesSubject = new Subject<IVehicle[]>();
    public apiVehicles$ : Observable<IVehicle[]> = this.apiVehiclesSubject.asObservable();

    private vehiclesSubject = new Subject<IVehicle[]>();
    public vehicles$ : Observable<IVehicle[]> = this.vehiclesSubject.asObservable();


    private apiPlanetsSubject = new Subject<IPlanet[]>();
    public apiPlanets$ : Observable<IPlanet[]> = this.apiPlanetsSubject.asObservable();

    private planetsSubject = new Subject<IPlanet[]>();
    public planets$ : Observable<IPlanet[]> = this.planetsSubject.asObservable();


    private totalTimeTakenSubject = new BehaviorSubject<number>(0);    
    public totalTimeTaken$ : Observable<number> = this.totalTimeTakenSubject.asObservable();

    private planetFoundOnSubject = new BehaviorSubject<string>('');
    public planetFoundOn$ : Observable<string> = this.planetFoundOnSubject.asObservable();
    
    private planetChangedSubject = new Subject<PlanetChange>(); 
    private planetChangedAction$ : Observable<PlanetChange> = this.planetChangedSubject.asObservable();
    public planetChanged(planetChange : PlanetChange) {
        this.planetChangedSubject.next(planetChange);
    }
    
    private vehicleChangedSubject = new Subject<VehicleChange>(); 
    private vehicleChangedAction$ : Observable<VehicleChange> = this.vehicleChangedSubject.asObservable();
    public vehicleChanged(vehicleChange : VehicleChange) {
        this.vehicleChangedSubject.next(vehicleChange);
    }
        

    public vehiclesUpdated$ = this.vehicleChangedAction$.pipe( 
        withLatestFrom(this.apiVehicles$, this.apiPlanets$, this.searchMap$),
        map(([vehicleChange, vehicles, planets, searchMap]) => {

            // update list of vehicles with updated available units
            const vehicleUpdates : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle} 
                        = this.getVehiclesWithUpdatedAvailableUnits(vehicles, vehicleChange, searchMap);
            const updatedVehicles : IVehicle[] = vehicleUpdates.updatedVehicles;
            this.updateVehicleInfo(updatedVehicles);       
            
            // update list of planets available to be searched after this change
            const updatedSearchMap = this.getUpdatedSearchMap(vehicleChange.widgetId, null, vehicleUpdates.lastUpdatedVehicle, searchMap);
            this.setSearchMap(updatedSearchMap);
            const planetsLeftForSearch = this.getPlanetsAvailableForSearch(planets, updatedSearchMap);                                    
            this.updatePlanetListForAvailability(planetsLeftForSearch);
            console.log(`vehiclesUpdated$ -> vehicleChangedAction$ withLatestFrom`,updatedSearchMap);
            return {widgetId : vehicleChange.widgetId, changer : 'vehicleUpdate'};
        })
    )     
    
    public planetsUpdated$ = this.planetChangedAction$
            .pipe(withLatestFrom(this.apiPlanets$, this.apiVehicles$, this.searchMap$),
                  map(([planetChange, planets, vehicles, searchMap]) => {

                        

                        // update list of planets available to be searched after this change
                        const updatedSearchMap = this.getUpdatedSearchMap(planetChange.widgetId, planetChange.newPlanet, null, searchMap);
                        this.setSearchMap(updatedSearchMap);
                        const planetsLeftForSearch = this.getPlanetsAvailableForSearch(planets, updatedSearchMap);                                    
                        this.updatePlanetListForAvailability(planetsLeftForSearch);

                        // update list of vehicles with updated available units
                        const vehicleUpdates : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle}
                                             = this.getVehiclesWithUpdatedAvailableUnits(vehicles, <VehicleChange>{ widgetId : planetChange.widgetId}, updatedSearchMap);
                        
                        const vehiclesWithUpdatedAvailableUnits: IVehicle[] = vehicleUpdates.updatedVehicles;
                        this.updateVehicleInfo(vehiclesWithUpdatedAvailableUnits);
                        console.log(`planetsUpdated$ -> withLatestFrom`,updatedSearchMap);
                        return {widgetId : planetChange.widgetId, changer : 'planetUpdate'};

                  })
            );

    
    private updateVehicleInfo(updatedVehicles: IVehicle[]): void {

        this.vehiclesSubject.next(updatedVehicles);
    }

    private getUpdatedSearchMap(changedWidgetId: number, changedWidgetPlanet: IPlanet, changedWidgetVehicle : IVehicle, searchMap: Map<number, ISearchAttempt>): Map<number, ISearchAttempt> {
        const newSearchMap = new Map<number, ISearchAttempt>();

        if(!changedWidgetPlanet) {
            changedWidgetPlanet = searchMap.get(changedWidgetId).searchedPlanet;
        }

        //const widgetId = planetChange.widgetId;
        // if there are any existing searchAttempts then
        // include searchAttempts to the left of currently changed widget         
        if(searchMap.size > 0) {            
           searchMap.forEach( (searchAttempt, widgetId) => {
               // this is the searchAttempt for widget to the left of the changed widget and it should be retained
                if(changedWidgetId > widgetId){

                    const unchangedSearchAttemptClone = { ...searchAttempt};
                    newSearchMap.set(widgetId, unchangedSearchAttemptClone);
                }                
           }); 
        }

        // this is the first planet being selected, simply add a new entry to the search map 
        if(changedWidgetPlanet && changedWidgetPlanet.name && changedWidgetPlanet.name !== 'Select'){
            newSearchMap.set(changedWidgetId,
                        <ISearchAttempt>{ 
                            searchedPlanet : {...changedWidgetPlanet},
                            vehicleUsed: changedWidgetVehicle
                            }
                        );
        }

        return newSearchMap;
    }              

    getVehiclesWithUpdatedAvailableUnits(vehicles : IVehicle[], vehicleChange: VehicleChange, searchMap: Map<number, ISearchAttempt>)
                : {updatedVehicles : IVehicle[], lastUpdatedVehicle: IVehicle} 
    {
    
        const updates = {updatedVehicles : null, lastUpdatedVehicle: null};
        let lastUpdatedVehicle : IVehicle;
        const updatedWidgetId = vehicleChange.widgetId;

        // Create a map which will store the updated available units of vehicles against vehicle name
        const vehiclesInUseMap = new Map<string, IVehicle>();
        for(let searchAttemptEntry of searchMap){
            // searchAttemptEntry is an array that has widgetId at index 0 and has serchAttempt at index 1
            const widgetId = searchAttemptEntry[0];
            const vehicleUsed = searchAttemptEntry[1].vehicleUsed; 

            // If current widget is to the left of updated widget
            // the available unit count for the vehicle used in that widget wont change
            if(widgetId < updatedWidgetId){
                if(vehicleUsed) {
                    lastUpdatedVehicle = {...vehicleUsed};
                    vehiclesInUseMap.set(vehicleUsed.name, lastUpdatedVehicle);
                }                
            }
            // We free up one unit from vehicle that was used earlier 
            // and decrease one unit from the newly selected vehicle.
            // If the control reaches here due to planet being updated in which case oldVehicle and newVehicle will be null 
            // then this widget slot will not have any vehicles
            else if(widgetId === updatedWidgetId){                
                    if(vehicleChange.oldVehicle && vehicleChange.oldVehicle.name) {
                        lastUpdatedVehicle = <IVehicle>{ ...vehicleChange.oldVehicle, availNumUnits : vehicleChange.oldVehicle.availNumUnits + 1};
                        vehiclesInUseMap.set(vehicleChange.oldVehicle.name, lastUpdatedVehicle);
                    }
                    if(vehicleChange.newVehicle && vehicleChange.newVehicle.name) {
                        lastUpdatedVehicle = <IVehicle>{ ...vehicleChange.newVehicle, availNumUnits : vehicleChange.newVehicle.availNumUnits - 1};
                        vehiclesInUseMap.set(vehicleChange.newVehicle.name, lastUpdatedVehicle);
                    }                
            }   
            // for all the widgets to the right of the updated widgets we can simply free up all the used units of the vehicles
            // and set availableUnits to totalUnits
            else if(widgetId > updatedWidgetId){
                lastUpdatedVehicle = <IVehicle>{ ...vehicleUsed, availNumUnits : vehicleUsed.totalNumUnits};
                vehiclesInUseMap.set(vehicleUsed.name, lastUpdatedVehicle);
            }
        }

        // Loop over the list of 
        const updatedVehicles : IVehicle[] = vehicles.map( vehicle => {
            const existingVehicle: IVehicle = vehiclesInUseMap.get(vehicle.name);
            
            if(existingVehicle){
                return {...existingVehicle};
            }
            else {
                return {...vehicle};
            }

        });

        
        return {updatedVehicles, lastUpdatedVehicle};
    }    

    private updatePlanetListForAvailability(planets : IPlanet[]) : void {

        this.planetsSubject.next(planets);
    }
    
    private getPlanetsAvailableForSearch(planets: IPlanet[], searchMap : Map<number, ISearchAttempt>)  : IPlanet[] {       
       
        if(searchMap.size > 0) {

            const planetsIncludedInSearch = new Set<string>();

            // create a set of all planets that have already been searched (i.e are present in searchMap)
            for(let searchAttempt of searchMap) {
                //searchAttempt is an array with index 0 being the widget id and index 1 being searchAttempt
                planetsIncludedInSearch.add(searchAttempt[1].searchedPlanet.name)
            }
        
            const planetsAvailableForSearch = planets.filter( planet => {                    
                    // if the planet is part of planetsIncludedInSearch set then filter it out                    
                    return !planetsIncludedInSearch.has(planet.name);
                });
        
            return [...planetsAvailableForSearch];
        }
        else {
            return [...planets];
        }
                        
    }           

    // todo remove, just for testing
    public some$ = combineLatest([this.planets$, this.planetChangedAction$])
    .pipe(
        map(([planets, planetChange]) => {
            
            console.log('combinelatest [this.planets$, this.planetChangedAction$]',planets, planetChange);
            return new PlanetUpdates(planets, planetChange, null);
            
        })
    )
    
    
    public isReadyForSearch$ = this.searchMap$.pipe(
        map(searchMap => {
            // if searchMap contains required entries then return true		
            if(searchMap && searchMap.size === this.MAX_SEARCH_ATTEMPTS_ALLOWED) {
                for(let entry of searchMap){
                    if(!entry[1] ||
                        (entry[1] && (!entry[1].searchedPlanet || !entry[1].vehicleUsed))) {
                        return false;
                    }
                }
                return true;
            }
            return false;
        })
    )        

    public initializeAppData() {

        forkJoin(
          this.vehicleService.getAllVehicles(),
          this.planetService.getAllPlanets(),
          this.finderService.getFalconFinderApiToken()
        )
        .pipe(
            catchError(
                error => {
                    this.setErrorMsg(error);
                    return EMPTY;
                }
            )
        )
        .subscribe( response => {
                      
            const vehicleList : IVehicle[] = response[0];
            const planetList : IPlanet[] = response[1];
            this.finderApiToken = response[2].token;                        

            this.apiVehiclesSubject.next(vehicleList);
            this.vehiclesSubject.next(vehicleList);

            this.apiPlanetsSubject.next(planetList);    
            this.planetsSubject.next(planetList);   
            
            // initialize/reset searchmap, error message and total time taken
            this.setSearchMap(new Map<number, ISearchAttempt>());
            this.errorMessageSubject.next('');
            this.totalTimeTakenSubject.next(0);
        },
          (error) => {
            this.setErrorMsg(error);
          }
        );

        this.searchMap$.pipe(
            tap(searchMap => {
                if(searchMap.size > 0){
                    let totalTimeTaken = 0;
                    for(let entry of searchMap){                        
                        let searchAttempt = entry[1];
                        if(searchAttempt.searchedPlanet && searchAttempt.searchedPlanet.name &&
                           searchAttempt.vehicleUsed && searchAttempt.vehicleUsed.name && searchAttempt.vehicleUsed.speed){

                                totalTimeTaken +=  (searchAttempt.searchedPlanet.distance / searchAttempt.vehicleUsed.speed);
                           }
                    }
                    this.totalTimeTakenSubject.next(totalTimeTaken);
                }
            }),
            distinctUntilChanged()
        ).subscribe( totalTime => {});

    }   
    
    private setSearchMap(searchMap: Map<number, ISearchAttempt>) : void {
        this.searchMapSubject.next(searchMap);
    }
        
    public findFalcon(request : IFindFalconRequest) {

        request.token = this.finderApiToken; 
        this.finderService.findFalcon(request)
           .subscribe( (response: IFindFalconResponse) => {
             let errorMsg = null;  
             if(response) {
   
               if(response.error) {
                 errorMsg = response.error;
                 this.setErrorMsg(errorMsg);
               }
               else if(response.status) {
   
                 if(response.status.trim().toLowerCase() === "success") {
   
                   if(response.planetName) {
                     this.planetFoundOnSubject.next(response.planetName);
                   }
                   else {
                     errorMsg = "Search api returned empty planet name";
                     this.setErrorMsg(errorMsg);
                   }
                 }
                 else if(response.status.trim().toLocaleLowerCase() === "false") {
                   errorMsg = "Failure! You were unable to find Falcone. Better luck next time.";
                   this.setErrorMsg(errorMsg);
                 }
                 else {
                   errorMsg = "Search api did not return a response status value."
                   this.setErrorMsg(errorMsg);
                 }
               }
             }
             else {
               errorMsg = "Search api returned invalid response."
               this.setErrorMsg(errorMsg);
             }
           },
           (error) => {
            this.setErrorMsg(error);
           }
           );
     }    

    private setErrorMsg(error) {
        this.errorMessageSubject.next(error);
    }

     resetApp(){
        this.router.navigate(['/finderboard/reset']);
      }      
}