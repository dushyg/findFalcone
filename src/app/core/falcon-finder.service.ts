import { Injectable } from "@angular/core";
import { HttpClient, HttpHeaders } from "@angular/common/http";
import { catchError, map } from "rxjs/operators";
import { handleError } from "./handleError";
import { Observable } from "rxjs";
import { IFindFalconeRequest } from "./models/findFalconeRequest";
import { IFindFalconeResponse } from "./models/findFalconeResponse";

@Injectable({
  providedIn: "root",
})
export class FalconFinderService {
  constructor(private http: HttpClient) {}

  private readonly falconFinderApiUrl =
    "https://findfalcone.herokuapp.com/find";

  public findFalcon(
    request: IFindFalconeRequest
  ): Observable<IFindFalconeResponse> {
    const httpOptions = {
      headers: new HttpHeaders({
        Accept: "application/json",
        "Content-Type": "application/json",
      }),
    };
    return this.http
      .post<IFindFalconeResponse>(this.falconFinderApiUrl, request, httpOptions)
      .pipe(
        map((response: any) => {
          return <IFindFalconeResponse>{
            planetName: response.planet_name,
            status: response.status,
            error: response.error,
          };
        }, catchError(handleError))
      );
  }
}
