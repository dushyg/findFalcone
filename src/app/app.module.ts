// tslint:disable: max-line-length
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { AppRoutingModule } from './app-routing.module';
import { ServiceWorkerModule } from '@angular/service-worker';
import { AppComponent } from './app.component';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { SharedModule } from './shared/shared.module';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import { environment } from 'src/environments/environment';
import { constants } from './shared/constants';
import {
  FinderBoardComponent,
  DestinationWidgetComponent,
  VehicleListComponent,
} from './finder-board/smartComponents';
import {
  DestinationWidgetListComponent,
  TypeaheadComponent,
} from './finder-board/presenterComponents';
import { AppData } from './finder-board/services/mockData/app.data';

@NgModule({
  declarations: [
    AppComponent,
    FinderBoardComponent,
    DestinationWidgetListComponent,
    DestinationWidgetComponent,
    VehicleListComponent,
    TypeaheadComponent,
  ],
  imports: [
    BrowserModule.withServerTransition({ appId: 'serverApp' }),
    AppRoutingModule,
    BrowserAnimationsModule,
    FormsModule,
    ReactiveFormsModule,
    SharedModule,
    HttpClientModule,
    environment.production
      ? []
      : HttpClientInMemoryWebApiModule.forRoot(AppData, {
          delay: 100,
          host: constants.apiDomain,
          apiBase: '/',
          rootPath: '/',
          passThruUnknownUrl: true,
        }),
    ServiceWorkerModule.register('ngsw-worker.js', {
      enabled: environment.production,
    }),
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
