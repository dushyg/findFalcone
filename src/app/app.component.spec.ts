import { TestBed, async, ComponentFixture, fakeAsync } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';
import { By } from '@angular/platform-browser';
import { FinderBoardShellComponent } from './features/falcon-finder/containers/finder-board-shell/finder-board-shell.component';
import { RouterOutlet } from '@angular/router';

describe('AppComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        AppComponent
      ],
    }).compileComponents();
  }));

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app).toBeTruthy();
  });

  it(`should have as title 'findingFalcone'`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;
    expect(app.title).toEqual('findingFalcone');
  });

  it('should contain app-finder-board-shell component', fakeAsync(() => {
    const fixture : ComponentFixture<AppComponent> = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const appComponentDebugElement = fixture.debugElement;            
    expect(appComponentDebugElement.query(By.directive(RouterOutlet))).toBeTruthy();        
  }));
});