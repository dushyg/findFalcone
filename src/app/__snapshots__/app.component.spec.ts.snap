// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the app 1`] = `
<app-root
  __ngContext__={[Function LRootView]}
  finderFacadeService={[Function FinderFacadeService]}
  title={[Function String]}
>
  <div
    class="falconApp"
  >
    <app-falcon-header>
      <nav
        class="headerContainer"
      >
        <div
          class="brandName"
        >
          <a
            class="brandAnchor"
            href="/finderboard"
          >
             Finding Falcone! 
          </a>
        </div>
        <div
          class="resetSection"
        >
          <a
            class="resetLink"
          >
             Reset 
          </a>
          <a
            class="geekTrustLink"
            href="https://www.geektrust.in"
            rel="noopener"
            target="_blank"
          >
             GeekTrust Home 
          </a>
          <a
            class="helpLink"
            href="#"
            routerlink="/"
          >
             Help 
          </a>
        </div>
      </nav>
    </app-falcon-header>
    <a
      class="skip-link"
      href="#appFalconBody"
    >
      Skip to main
    </a>
    <div
      class="appFalconBody"
      id="appFalconBody"
    >
      
      
      <router-outlet />
    </div>
    <app-falcon-footer>
      <nav
        class="footer"
      >
        <div
          class="footer-container"
        >
          <div
            class="geekTrustProblem"
          >
             Coding Problem - www.geektrust.in 
          </div>
        </div>
      </nav>
    </app-falcon-footer>
  </div>
</app-root>
`;
